### PR TITLE
Avoid duplication of effort

### PR DESCRIPTION
**Pull request by ethanc8**
I don't actually intend for this pull request to be merged (I wanted to file an issue but this repo doesn't accept issues), but we should avoid duplication of effort in our CoreFoundation ports. [My CoreFoundation port](https://github.com/ethanc8/CoreFoundation) is based off of swift-corelibs-foundation from Catalina (at some version of Swift 5.3, I don't remember which and I'm not very good at Git), and it supports bridging with GNUstep-base. In my `gnustep` branch, I just have a plain port of CoreFoundation to be buildable with GNUstep Make. Could you please explain what changes you've made in your fork more broadly and whether it would be better to build off of my fork or to start a new fork from a newer version of CoreFoundation?

### CODE REVIEW

**Comment by ethanc8** [comment]
@trunkmaster What do you think?

**Comment by trunkmaster** [comment]
@ethanc8 hi, probably I've missed your 2-week-ago PR, sorry.
> I don't actually intend for this pull request to be merged (I wanted to file an issue but this repo doesn't accept issues), but we should avoid duplication of effort in our CoreFoundation ports. 

It's excellent idea to merge our efforts! I've created this repo to avoid keeping patches inside [NEXTSPACE](https://github.com/trunkmaster/nextspace).

> [My CoreFoundation port](https://github.com/ethanc8/CoreFoundation) is based off of swift-corelibs-foundation from Catalina (at some version of Swift 5.3, I don't remember which and I'm not very good at Git), and it supports bridging with GNUstep-base.

My repo is based on the 5.9.2 branch of original Swift's version. I've noticed that CoreFoundation in Swift's repository changes not that often. I think your bridging to GNUstep-base would be a killer feature!

> In my `gnustep` branch, I just have a plain port of CoreFoundation to be buildable with GNUstep Make. Could you please explain what changes you've made in your fork more broadly and whether it would be better to build off of my fork or to start a new fork from a newer version of CoreFoundation?

The main idea behind this fork is to create CoreFoundation to write C-based programms with all the power of CoreFoundation. For example, it helped me to rewrote large part of WindowMaker and get rid of WINGs and data management part of it (preferences, arrays, dictionaries etc.).

The next major part is CFRunLoop and CFFileDescriptor implementation. I've managed to get X11 messages by monitoring CFFileDescriptor of X server connection inside CFRunLoop. 

Next big deal is a CFNotificationCenter implementation: I use it inside WM to pass messages inside C code and create a NSNotificationCenter<->CFNotificationCenter bridge. This part is usefull for notifications exchange between Obj-C and C programms. For example, GNUstep applications may send notifications to global notifications center and be catched by window manager (CFRunLoop, CFNotificationCenter) and vice versa.

So to sum up, i've made the following changes:
- made Swift's version of CoreFoundation to build as shared library;
- added missing CFFileDescriptor implementation - testing with NEXTSPACE;
- added missing CFNotificationCenter implementation - testing with NEXTSPACE;
- removed all Swift-specific code - in this repo I have only CoreFoundation subdirectory.

You could track all my changes starting from https://github.com/trunkmaster/apple-corefoundation/commit/03eea750d17856310a5a166daecc05fad8be067b commit.

If you want I can configure this repo available for issues and PR. Also I want to keep this repo as CMake-based project. Do you want to become a contributor or will add your changes as PRs?

**Comment by trunkmaster** [comment]
@ethanc8 I don't quite understand what's going on here. I see CoreFoundation subdirectory with a bunch of C files (.c) renamed into ObjC files (.m). What do you include Swift code? How it should correspond to existing CF sources that already exists in my CF source tree?

Could you please explain your idea of implementing CoreFoundation to GNUstep Base bridge? What mechanics of bridging you assume to implement? How ObjC classes should gain access to CF types and vice versa?

**Comment by ethanc8** [comment]
@trunkmaster Please see https://github.com/ethanc8/CoreFoundation/blob/gnustep-bridging/CoreFoundation/Documentation/Bridging.md and https://github.com/ethanc8/CoreFoundation/blob/gnustep-bridging/CoreFoundation/TODO.md. I will try to write a better explanation of my changes this week.

**Comment by ethanc8** [comment]
(this is a simplification, you need to read my commits in order to know everything I did, as I forgot many things that I did two years ago)
* I branched off of  [swift-5.3-RELEASE](https://github.com/apple/swift-corelibs-foundation/tree/swift-5.3-RELEASE), based on the [Catalina source code](https://github.com/apple/swift-corelibs-foundation/pull/2787), with [release notes here](https://github.com/apple/swift-corelibs-foundation/pull/2782) (CoreFoundation versions 1665.15~1677.104)
* I renamed all `.c` files to `.m`.
* There were various changes I did to make it build with GNUstep Make -- this was a couple of years ago and I don't remember exactly
* The deployment runtime can be `DEPLOYMENT_RUNTIME_OBJC` or `DEPLOYMENT_RUNTIME_SWIFT`. If `DEPLOYMENT_RUNTIME_OBJC`, it will bridge to Darwin Foundation.framework, and if `DEPLOYMENT_RUNTIME_SWIFT`, it will bridge to the Windows/Linux Swift runtime. I added a `DEPLOYMENT_RUNTIME_GNUSTEP_LIBOBJC2`, and added the correct behavior for that case to everywhere that did a `#ifdef DEPLOYMENT_RUNTIME_OBJC`.
* I modified `CF_OBJC_FUNCDISPATCHV` and other macros which were related to bridging, to work on Objective-C.
* I implemented a lot of toll-free bridged classes according to https://github.com/ethanc8/CoreFoundation/blob/gnustep-bridging/CoreFoundation/Documentation/Bridging.md, and based on the WinObjC Foundation.

**Comment by trunkmaster** [comment]
> (this is a simplification, you need to read my commits in order to know everything I did, as I forgot many things that I did two years ago)
> 
> * I branched off of  [swift-5.3-RELEASE](https://github.com/apple/swift-corelibs-foundation/tree/swift-5.3-RELEASE), based on the [Catalina source code](https://github.com/apple/swift-corelibs-foundation/pull/2787), with [release notes here](https://github.com/apple/swift-corelibs-foundation/pull/2782) (CoreFoundation versions 1665.15~1677.104)

OK, my version of CoreFoundation is newer - 5.9.3.

> * I renamed all `.c` files to `.m`.

What for? Is there any reason to make C library to look like ObjC?

> * There were various changes I did to make it build with GNUstep Make -- this was a couple of years ago and I don't remember exactly

I see no reason to build Apple library with GNUstep Make - my version builds before GNUstep Make even installed. Do you?

> * The deployment runtime can be `DEPLOYMENT_RUNTIME_OBJC` or `DEPLOYMENT_RUNTIME_SWIFT`. If `DEPLOYMENT_RUNTIME_OBJC`, it will bridge to Darwin Foundation.framework, and if `DEPLOYMENT_RUNTIME_SWIFT`, it will bridge to the Windows/Linux Swift runtime. I added a `DEPLOYMENT_RUNTIME_GNUSTEP_LIBOBJC2`, and added the correct behavior for that case to everywhere that did a `#ifdef DEPLOYMENT_RUNTIME_OBJC`.
>
> * I modified `CF_OBJC_FUNCDISPATCHV` and other macros which were related to bridging, to work on Objective-C.

Great.That's the most interesting part for me.

Could you please extract those 2 block of changes and prepare it to apply to my source tree? See my first post in this discussion. 
At this point all of these changes couldn't be aplied to trunkmaster/apple-corefoundation.

> * I implemented a lot of toll-free bridged classes according to https://github.com/ethanc8/CoreFoundation/blob/gnustep-bridging/CoreFoundation/Documentation/Bridging.md, and based on the WinObjC Foundation.

Thanks, I'll try to understand idea.

**Comment by ethanc8** [comment]
> What for? Is there any reason to make C library to look like ObjC?

The point is not to look like ObjC; the point is to allow CoreFoundation to send Objective-C messages without having to manually call `objc_msgSend` (which is rather error-prone). `CF_OBJC_FUNCDISPATCHV` in my version simmply sends Objective-C messages.

> I see no reason to build Apple library with GNUstep Make - my version builds before GNUstep Make even installed. Do you?

I know how to use GNUstep Make, and don't know how to use CMake. GNUstep Make is also good at dealing with GNUstep libraries. That's pretty much the only reason why.

> Could you please extract those 2 block of changes and prepare it to apply to my source tree? See my first post in this discussion.

Sorry, I don't want to work on this. I want to focus on getting the tests to pass; if you need a CoreFoundation with bridging, just use mine. If I do want to rebase on a newer CoreFoundation, I will rebase on Ventura CoreFoundation, not Monterey CoreFoundation.
